### PR TITLE
feat: Preact Adapter

### PR DESCRIPTION
**Pull request by theVedanta**
## 🎯 Changes

Makes a Preact Adapter for Tanstack Query. The main motivations are bundle sizes and ensuring compatibility with the events architecture. Docs and Tests (which were the major effort here) have also been written.

## ✅ Checklist

- [x] I have followed the steps in the [Contributing guide](https://github.com/TanStack/query/blob/main/CONTRIBUTING.md).
- [x] I have tested this code locally with `pnpm run test:pr`.

## 🚀 Release Impact

- [ ] This change affects published code, and I have generated a [changeset](https://github.com/changesets/changesets/blob/main/docs/adding-a-changeset.md).
- [ ] This change is docs/CI/dev-only (no release).

## Summary by CodeRabbit

* **New Features**
  * Launches a Preact-first query package: hooks, providers, hydration & error-reset boundaries, Suspense/infinite queries, mutations, prefetch utilities, and a consolidated public API.

* **Documentation**
  * Extensive Preact guides and full API reference pages with examples and TypeScript docs.

* **Tests**
  * Large suite of unit and TypeScript tests covering hooks, SSR/hydration, suspense, infinite queries, mutations, persisters and edge cases.

* **Examples & Tooling**
  * New Preact example app and Vite/TS/ESLint setup.

* **Chores**
  * Build, packaging, and test setup updates; changelog entries.

### CODE REVIEW

File: packages/preact-query/src/errorBoundaryUtils.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
+'use client'
```
Comment (rschristian) [inline packages/preact-query/src/errorBoundaryUtils.ts]: FWIW, no need for these as they do nothing in Preact. None of our tooling acknowledges these directives.

File: packages/preact-query/src/HydrationBoundary.tsx [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+export interface HydrationBoundaryProps {
+  state: DehydratedState | null | undefined
+  options?: OmitKeyof<HydrateOptions, 'defaultOptions'> & {
+    defaultOptions?: OmitKeyof<
+      Exclude<HydrateOptions['defaultOptions'], undefined>,
+      'mutations'
+    >
+  }
+  children?: React.ReactNode
```
Comment (rschristian) [inline packages/preact-query/src/HydrationBoundary.tsx]: You'd want `ComponentChildren` from `preact`

File: packages/preact-query/src/errorBoundaryUtils.ts (hunk 1 above)
Comment (theVedanta) [reply packages/preact-query/src/errorBoundaryUtils.ts]: So sorry I missed these

File: packages/preact-query/src/HydrationBoundary.tsx (hunk 2 above)
Comment (theVedanta) [reply packages/preact-query/src/HydrationBoundary.tsx]: Fixed!

File: packages/preact-query/src/errorBoundaryUtils.ts (hunk 1 above)
Comment (rschristian) [reply packages/preact-query/src/errorBoundaryUtils.ts]: It's certainly not a problem, no need for an apology. Just wanted to drop mention in case you weren't aware is all.

**Comment by alextompkins** [comment]
Just wanted to note my appreciation for this, as a member of a team using Tanstack Query with preact 🫶

**Comment by theVedanta** [comment]
> Just wanted to note my appreciation for this, as a member of a team using Tanstack Query with preact 🫶 

thank you 😭❤️
feel free to let me know any thoughts you might have on reducing bundle size \:)

File: packages/preact-query/src/errorBoundaryUtils.ts (hunk 1 above)
Comment (theVedanta) [reply packages/preact-query/src/errorBoundaryUtils.ts]: yo @rschristian I had some questions and would be great to get a fresh set of eyes on some of these tests. Can I reach out to you on discord or any other platform? totally cool if not; I understand it's holiday season

Comment (rschristian) [reply packages/preact-query/src/errorBoundaryUtils.ts]: I don't use Discord, but you can DM me on Twitter or Bluesky (both linked from my profile here) or the [Preact Slack](https://join.slack.com/t/preact/shared_invite/zt-3jn7lkdar-0ZkZpo889rl6p2Oyrmoo1w), whichever is convenient. Will answer when I can

Comment (theVedanta) [reply packages/preact-query/src/errorBoundaryUtils.ts]: texted on Slack!

File: packages/preact-query/src/__tests__/ErrorBoundary/ErrorBoundary.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+      this.setState(initialState)
+    }
+  }
+
+  render() {
+    const { children, fallbackRender, FallbackComponent, fallback } = this.props
+    const { didCatch, error } = this.state
+
+    let childToRender = children
```
Comment (rschristian) [inline packages/preact-query/src/__tests__/ErrorBoundary/ErrorBoundary.ts:92]: ~~I really wouldn't recommend this pattern, it'll cause a heck of a lot of thrashing in most apps.~~

~~Also, unlike React, we do have a [`useErrorBoundary` hook](https://preactjs.com/guide/v10/hooks#useerrorboundary) which you may wish to use instead.~~

Edit: Missed this was in the `__tests__` directory, oops. Still likely unnecessary, you can just create a tiny wrapper like I showed in Slack.

Comment (theVedanta) [reply packages/preact-query/src/__tests__/ErrorBoundary/ErrorBoundary.ts:92]: oh haha yes I'm planning on using that wrapper you sent lmao. Sorry this just ended up here because of some other tests already using it. Will remove the unnecessary parts soon :)

File: packages/preact-query/package.json [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    "@tanstack/query-test-utils": "workspace:*",
+    "@testing-library/preact": "^3.2.4",
+    "@testing-library/react-render-stream": "^2.0.0",
+    "cpy-cli": "^5.0.0",
+    "eslint-config-preact": "^2.0.0",
+    "npm-run-all2": "^5.0.0",
+    "preact": "^10.28.0",
+    "preact-iso": "^2.11.1",
+    "preact-render-to-string": "^6.6.4",
+    "typescript-eslint": "^8.50.0"
```
Comment (TkDodo) [inline packages/preact-query/package.json]: we get typescript-eslint from the tanstack-config package, and it’s included in the root config, so I don’t think we should need it to be added explicitly here. We don’t have this for the other packages either but we do have typed linting!

File: packages/preact-query/root.eslint.config.js [hunk 5]
```diff
@@ -0,0 +1 @@
+../../eslint.config.js
```
[the remark this answers is not in the document]
Comment (TkDodo) [reply packages/preact-query/root.eslint.config.js:1]: it’s a symlink!

File: packages/preact-query/package.json (hunk 4 above)
Comment (theVedanta) [reply packages/preact-query/package.json]: oh really? I'm sorry I didn't realize.

File: docs/framework/preact/reference/functions/HydrationBoundary.md [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+title: HydrationBoundary
+---
+
+# Function: HydrationBoundary()
+
+```ts
+function HydrationBoundary(__namedParameters): ComponentChildren;
+```
+
+Defined in: [preact-query/src/HydrationBoundary.tsx:24](https://github.com/theVedanta/query/blob/main/packages/preact-query/src/HydrationBoundary.tsx#L24)
```
Comment (TkDodo) [inline docs/framework/preact/reference/functions/HydrationBoundary.md:12]: I’ve set up correct docs linking here:
[aa88156](https://github.com/TanStack/query/pull/9935/commits/aa8815665fe898c820d4dec3290755c7e1c2c0dd)

some things:
- If we do links to github, they should point to the `query` repo.
- We likely shouldn’t point to `main` but use permalinks to a specific commit hash to avoid pointing into the void when line numbers or file names change
- speaking of getting outdated, unless this is auto-generated, it feels like it can get outdated really fast. We don’t have these pointers to github in the other  adapters

File: docs/framework/preact/reference/type-aliases/AnyUseBaseQueryOptions.md
Comment (TkDodo) [inline docs/framework/preact/reference/type-aliases/AnyUseBaseQueryOptions.md:1]: are these auto-generated? Some of those types are probably exported by accident in the react adapter too because we `export * from './types.ts' :/

we usually say that the public interface is whatever we document. If we now auto generate these, everything becomes public interface forever.

It would be great if you could align the docs with whatever we have for the other adapters.

File: scripts/generate-docs.ts
Comment (TkDodo) [inline scripts/generate-docs.ts:1]: I legit didn’t know we had this 😂.

ignore my comments about the docs then 🤣

Comment (theVedanta) [reply scripts/generate-docs.ts:1]: lmao happens to me sometimes too

what comments 🤣🤣

Comment (TkDodo) [reply scripts/generate-docs.ts:1]: I’m trying to find out why CI is failing, it must be because of some changes in the lockfile I guess :/

Comment (theVedanta) [reply scripts/generate-docs.ts:1]: yep, definitely a lockfile problem. I've fixed the lock but I should let you know I'm seeing some `test:eslint` errors when I run the test suite. 
Usually this kind:
```
Avoid calling `expect` inside conditional statements (eslint vitest/no-conditional-expect)
```

Comment (TkDodo) [reply scripts/generate-docs.ts:1]: github says there are still conflicts with the main lockfile

Comment (theVedanta) [reply scripts/generate-docs.ts:1]: lemme check them out

Comment (theVedanta) [reply scripts/generate-docs.ts:1]: oh I see I moved the jj bookmark wrong. Should be good now.
